src/kakarot/instructions/block_information.cairo

```diff
@@ -32,6 +34,62 @@ namespace BlockInformation {
     const GAS_COST_SELFBALANCE = 5;
     const GAS_COST_BASEFEE = 2;
 
+    // @notice COINBASE operation.
+    // @dev Get the hash of one of the 256 most recent complete blocks.
+    // @dev 0 if the block number is not in the valid range.
+    // @custom:since Frontier
+    // @custom:group Block Information
+    // @custom:gas 20
+    // @custom:stack_consumed_elements 1
+    // @custom:stack_produced_elements 1
+    // @param ctx The pointer to the execution context
+    // @return The pointer to the updated execution context.
+    func exec_blockhash{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(ctx: model.ExecutionContext*) -> model.ExecutionContext* {
+        alloc_locals;
+        // Get the blockNumber
+        let (stack: model.Stack*, _block_number: Uint256) = Stack.pop(ctx.stack);
+        let block_number: felt = Helpers.uint256_to_felt(val=_block_number);
+
+        // Check if blockNumber is within bounds by checking with current block number
+        // Valid range is the last 256 blocks (not including the current one)
+        let (local current_block_number: felt) = get_block_number();
+        let in_range = is_in_range(block_number, current_block_number - 256, current_block_number);
+
+        // If not in range, return 0
+        if (in_range == FALSE) {
+            let blockhash: Uint256 = Helpers.to_uint256(val=0);
+            let stack: model.Stack* = Stack.push(self=stack, element=blockhash);
+
+            // Update the execution context.
+            // Update context stack.
+            let ctx = ExecutionContext.update_stack(self=ctx, new_stack=stack);
+            // Increment gas used.
+            let ctx = ExecutionContext.increment_gas_used(self=ctx, inc_value=GAS_COST_BLOCKHASH);
+            return ctx;
+        }
+
+        // Get blockhash from corresponding block number from registry
+        // Then push it to stack
+        let (blockhash_registry_address_: felt) = blockhash_registry_address.read();
+        let (blockhash_: felt) = IBlockhashRegistry.get_blockhash(
+            contract_address=blockhash_registry_address_, block_number=_block_number
+        );
+        let blockhash: Uint256 = Helpers.to_uint256(val=blockhash_);
+        let stack: model.Stack* = Stack.push(self=stack, element=blockhash);
+
+        // Update the execution context.
+        // Update context stack.
+        let ctx = ExecutionContext.update_stack(self=ctx, new_stack=stack);
+        // Increment gas used.
+        let ctx = ExecutionContext.increment_gas_used(self=ctx, inc_value=GAS_COST_BLOCKHASH);
+        return ctx;
```

Comment:
This is the BLOCKHASH opcode implementation. Takes the block number from the stack and checks if it is within valid range. If it is, Kakarot uses it to call the blockhash registry and retrieve the blockhash. Due to revoked references, I had to update the stack and execution context like this